File: src/main/java/org/lice/api/scripting/LiceContext.java
package org.lice.api.scripting;

import javax.script.Bindings;
import javax.script.ScriptContext;
import java.io.Reader;
import java.io.Writer;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Created by Glavo on 17-9-22.
 *
 * @author Glavo
 * @since 4.0.0
 */
public final class LiceContext implements ScriptContext {
	private Map<Integer, Bindings> allBindings;

	public LiceContext() {
		this.allBindings = new HashMap<>();
	}


	@Override
	public void setBindings(Bindings bindings, int scope) {
		this.allBindings.put(scope, bindings);
	}

	@Override
	public Bindings getBindings(int scope) {
		return allBindings.get(scope);
	}

	@Override
	public void setAttribute(String name, Object value, int scope) {
		Bindings bindings = allBindings.get(scope);
		Objects.requireNonNull(bindings);
		bindings.put(name, value);
	}

	@Override
	public Object getAttribute(String name, int scope) {
		Bindings bindings = allBindings.get(scope);
		Objects.requireNonNull(bindings);
		return bindings.get(name);
	}

	@Override
	public Object removeAttribute(String name, int scope) {
		return null;
	}

	@Override
	public Object getAttribute(String name) {
		return null;
	}

	@Override
	public int getAttributesScope(String name) {
		return 0;
	}

	@Override
	public Writer getWriter() {
		return null;
	}

	@Override
	public Writer getErrorWriter() {
		return null;
	}

	@Override
	public void setWriter(Writer writer) {

	}

	@Override
	public void setErrorWriter(Writer writer) {

	}

	@Override
	public Reader getReader() {
		return null;
	}

	@Override
	public void setReader(Reader reader) {

	}

	@Override
	public List<Integer> getScopes() {
		return null;
	}
}

File: src/main/kotlin/org/lice/api/scripting/LiceContext.kt
package org.lice.api.scripting

import org.lice.core.SymbolList
import org.lice.model.Node
import org.lice.model.ValueNode
import org.lice.util.cast
import java.io.*
import javax.script.Bindings
import javax.script.ScriptContext

/**
 * Created by Glavo on 17-6-8.
 *
 * @author Glavo
 * @since 0.1.0
 */

class LiceContext
@JvmOverloads
constructor(
		var bindings: SymbolList = SymbolList(),
		private var reader: Reader = InputStreamReader(System.`in`),
		private var writer: Writer = OutputStreamWriter(System.out),
		private var errorWriter: Writer = OutputStreamWriter(System.err)
) : ScriptContext {
	override fun getReader(): Reader = reader

	override fun setReader(reader: Reader) {
		this.reader = reader
	}

	override fun getWriter(): Writer = writer

	override fun setWriter(writer: Writer?) {
		this.writer = writer!!
	}

	fun removeAttribute(name: String) = bindings.removeVariable(name)
	override fun removeAttribute(name: String, scope: Int) = removeAttribute(name)

	override fun getBindings(scope: Int): Bindings = bindings

	override fun setBindings(bindings: Bindings?, scope: Int) {
		this.bindings = cast(bindings)
	}

	override fun getErrorWriter(): Writer = errorWriter

	override fun setErrorWriter(writer: Writer) {
		this.errorWriter = writer
	}

	override fun getAttribute(name: String) = (bindings.getVariable(name) as Node).eval()

	fun setAttribute(name: String, value: Any?) {
		bindings.defineVariable(name, ValueNode(value))
	}

	override fun getAttribute(name: String, scope: Int) = getAttribute(name)
	override fun setAttribute(name: String, value: Any?, scope: Int) = setAttribute(name, value)

	override fun getScopes(): MutableList<Int> = throw UnsupportedOperationException()
	override fun getAttributesScope(name: String?): Int = throw UnsupportedOperationException()
}